File: LC/Medium/309.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        sold = -inf
        held = -inf
        reset = 0
        
        for price in prices:
            sold, held, reset = held + price, max(held, reset - price), max(reset, sold)
            
        return max(sold, reset)

File: LC/C++/Medium/309.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int sold = std::numeric_limits<int>::min();
        int held = std::numeric_limits<int>::min();
        int reset = 0;

        for (int price : prices) {
            int temp = sold;
            sold = held + price;
            held = max(held, reset - price);
            reset = max(reset, temp);
        }

        return std::max(sold, reset);
    }
};